File: content/tracks/dsa-typescript/concepts/12-heaps-and-priority-queues/heaps-and-priority-queues_test.ts
import { MinHeap, kLargestElements, mergeKSortedLists, ListNode } from "./stub";

let passed = 0;
let failed = 0;

function test(name: string, fn: () => void): void {
  try {
    fn();
    console.log(`  PASS: ${name}`);
    passed++;
  } catch (e: unknown) {
    const msg = e instanceof Error ? e.message : String(e);
    console.log(`  FAIL: ${name} — ${msg}`);
    failed++;
  }
}

function assertEqual<T>(actual: T, expected: T): void {
  const a = JSON.stringify(actual);
  const b = JSON.stringify(expected);
  if (a !== b) throw new Error(`expected ${b}, got ${a}`);
}

function makeList(vals: number[]): ListNode | null {
  if (vals.length === 0) return null;
  const head = new ListNode(vals[0]);
  let curr = head;
  for (let i = 1; i < vals.length; i++) {
    curr.next = new ListNode(vals[i]);
    curr = curr.next;
  }
  return head;
}

function toArray(head: ListNode | null): number[] {
  const result: number[] = [];
  let curr = head;
  while (curr !== null) { result.push(curr.val); curr = curr.next; }
  return result;
}

// --- MinHeap ---

test("MinHeap: pop returns minimum", () => {
  const h = new MinHeap<number>((a, b) => a - b);
  h.push(3); h.push(1); h.push(2);
  assertEqual(h.pop(), 1);
});

test("MinHeap: peek without removing", () => {
  const h = new MinHeap<number>((a, b) => a - b);
  h.push(5); h.push(2); h.push(8);
  assertEqual(h.peek(), 2);
  assertEqual(h.size(), 3);
});

test("MinHeap: sorted extraction order", () => {
  const h = new MinHeap<number>((a, b) => a - b);
  [5, 3, 8, 1, 4].forEach((n) => h.push(n));
  const order: number[] = [];
  while (h.size() > 0) order.push(h.pop()!);
  assertEqual(order, [1, 3, 4, 5, 8]);
});

test("MinHeap: empty pop returns undefined", () => {
  const h = new MinHeap<number>((a, b) => a - b);
  assertEqual(h.pop(), undefined);
});

test("MinHeap: single element push then pop", () => {
  const h = new MinHeap<number>((a, b) => a - b);
  h.push(42);
  assertEqual(h.pop(), 42);
  assertEqual(h.size(), 0);
});

test("MinHeap: max-heap via reversed comparator", () => {
  const h = new MinHeap<number>((a, b) => b - a);
  [3, 1, 4, 1, 5].forEach((n) => h.push(n));
  assertEqual(h.pop(), 5);
});

test("MinHeap: size tracks correctly", () => {
  const h = new MinHeap<number>((a, b) => a - b);
  assertEqual(h.size(), 0);
  h.push(1); h.push(2);
  assertEqual(h.size(), 2);
  h.pop();
  assertEqual(h.size(), 1);
});

// --- kLargestElements ---

test("kLargestElements: k=2", () => {
  const result = kLargestElements([3, 2, 1, 5, 6, 4], 2).sort((a, b) => a - b);
  assertEqual(result, [5, 6]);
});

test("kLargestElements: k=1", () => {
  const result = kLargestElements([3, 2, 1, 5, 6, 4], 1);
  assertEqual(result, [6]);
});

test("kLargestElements: k equals length", () => {
  const result = kLargestElements([1, 2, 3], 3).sort((a, b) => a - b);
  assertEqual(result, [1, 2, 3]);
});

test("kLargestElements: negative numbers", () => {
  const result = kLargestElements([-1, -5, -3, -2], 2).sort((a, b) => a - b);
  assertEqual(result, [-2, -1]);
});

test("kLargestElements: duplicates", () => {
  const result = kLargestElements([2, 2, 2, 3, 1], 3).sort((a, b) => a - b);
  assertEqual(result, [2, 2, 3]);
});

// --- mergeKSortedLists ---

test("mergeKSortedLists: three lists", () => {
  const lists = [makeList([1, 4, 5]), makeList([1, 3, 4]), makeList([2, 6])];
  assertEqual(toArray(mergeKSortedLists(lists)), [1, 1, 2, 3, 4, 4, 5, 6]);
});

test("mergeKSortedLists: empty input", () => {
  assertEqual(mergeKSortedLists([]), null);
});

test("mergeKSortedLists: single list", () => {
  assertEqual(toArray(mergeKSortedLists([makeList([1, 2, 3])])), [1, 2, 3]);
});

test("mergeKSortedLists: some lists are null", () => {
  const lists = [null, makeList([1, 3]), null, makeList([2, 4])];
  assertEqual(toArray(mergeKSortedLists(lists)), [1, 2, 3, 4]);
});

test("mergeKSortedLists: all single elements", () => {
  const lists = [makeList([3]), makeList([1]), makeList([2])];
  assertEqual(toArray(mergeKSortedLists(lists)), [1, 2, 3]);
});

console.log(`\n${passed} passed, ${failed} failed`);
if (failed > 0) process.exit(1);

File: content/tracks/dsa-python/concepts/12-heaps-and-priority-queues/heaps-and-priority-queues_test.py
import sys
from stub import MinHeap, k_largest_elements, merge_k_sorted_lists, ListNode, _make_list, _to_vals

passed = 0
failed = 0


def test(name: str, condition: bool, detail: str = "") -> None:
    global passed, failed
    if condition:
        print(f"  PASS: {name}")
        passed += 1
    else:
        msg = f" — {detail}" if detail else ""
        print(f"  FAIL: {name}{msg}")
        failed += 1


# MinHeap
h = MinHeap()
h.push(5); h.push(3); h.push(8)
test("heap: peek after 3 pushes", h.peek() == 3, f"got {h.peek()}")
test("heap: len after 3 pushes", len(h) == 3, f"got {len(h)}")
popped = h.pop()
test("heap: pop returns min", popped == 3, f"got {popped}")
test("heap: peek after pop", h.peek() == 5, f"got {h.peek()}")
test("heap: len after pop", len(h) == 2, f"got {len(h)}")
h2 = MinHeap()
h2.push(1)
test("heap: single element peek", h2.peek() == 1)
test("heap: single element pop", h2.pop() == 1)
test("heap: empty after pop", len(h2) == 0)
h3 = MinHeap()
for v in [10, 4, 15, 1, 7]:
    h3.push(v)
pops = [h3.pop() for _ in range(5)]
test("heap: pops in sorted order", pops == [1, 4, 7, 10, 15], f"got {pops}")

# k_largest_elements
r = sorted(k_largest_elements([3, 2, 1, 5, 6, 4], 2))
test("k_largest: k=2", r == [5, 6], f"got {r}")
r = sorted(k_largest_elements([3, 2, 3, 1, 2, 4, 5, 5, 6], 4))
test("k_largest: k=4 with duplicates", r == [4, 5, 5, 6], f"got {r}")
r = sorted(k_largest_elements([1], 1))
test("k_largest: single element", r == [1], f"got {r}")
r = sorted(k_largest_elements([1, 2, 3, 4, 5], 5))
test("k_largest: k=all elements", r == [1, 2, 3, 4, 5], f"got {r}")
r = sorted(k_largest_elements([-1, -2, -3, -4], 2))
test("k_largest: negative numbers", r == [-2, -1], f"got {r}")

# merge_k_sorted_lists
r = _to_vals(merge_k_sorted_lists([_make_list([1, 4, 5]), _make_list([1, 3, 4]), _make_list([2, 6])]))
test("merge_k: standard 3 lists", r == [1, 1, 2, 3, 4, 4, 5, 6], f"got {r}")
r = _to_vals(merge_k_sorted_lists([]))
test("merge_k: empty list of lists", r == [], f"got {r}")
r = _to_vals(merge_k_sorted_lists([None]))
test("merge_k: single empty list", r == [], f"got {r}")
r = _to_vals(merge_k_sorted_lists([_make_list([1])]))
test("merge_k: single list single element", r == [1], f"got {r}")
r = _to_vals(merge_k_sorted_lists([_make_list([1, 2, 3])]))
test("merge_k: single list", r == [1, 2, 3], f"got {r}")

print(f"\n{passed} passed, {failed} failed")
if failed > 0:
    sys.exit(1)
